src/mesh.cc

```diff
@@ -1,96 +1,84 @@
-void pipenetwork::Mesh::create_nodes(
-    const std::vector<Eigen::Vector3d>& coords) {
-  Index index = 0;
-  for (const auto& coord : coords) {
-    nodes_.emplace(std::pair<Index, std::shared_ptr<pipenetwork::Node>>(
-        index, std::make_shared<pipenetwork::Node>(index, coord)));
-    ++index;
+  bool stat{false};
+  Index i = 0;
+  for (const auto& id : ids) {
+    nodes_.emplace(id, std::make_shared<pipenetwork::Junction>(
+                           id, elevations[i], demands[i], leak_diameters[i]));
+    ++i;
   }
+  return stat;
+}
+
+bool pipenetwork::Mesh::create_reservoirs(const std::vector<Index>& ids,
+                                          const std::vector<double>& heads) {
+  Index i = 0;
+  bool stat{false};
+  for (const auto& id : ids) {
+    nodes_.emplace(id, std::make_shared<pipenetwork::Reservoir>(id, heads[i]));
+    ++i;
+  }
+  stat = true;
+  return stat;
 }
 
-// Create pipe pointers and assign indices based on the nodes at its ends
 bool pipenetwork::Mesh::create_pipes(
+    const std::vector<Index>& ids,
     const std::vector<std::pair<Index, Index>>& nodeids,
-    const std::vector<double>& diameter, const std::vector<double>& roughness,
-    const std::vector<bool>& pipe_status) {
-  bool status = true;
-  Index index = 0;
+    const std::vector<double>& length, const std::vector<double>& diameter,
+    const std::vector<double>& roughness,
+    const std::vector<Pipe_status>& status) {
+
+  Index i = 0;
+  bool stat{false};
   try {
-    for (const auto& nodeid : nodeids) {
-      std::array<std::shared_ptr<pipenetwork::Node>, 2> nodes;
-      nodes.at(0) = nodes_.at(nodeid.first);
-      nodes.at(1) = nodes_.at(nodeid.second);
-      pipes_.emplace(std::pair<Index, std::shared_ptr<pipenetwork::Pipe>>(
-          index, std::make_shared<pipenetwork::Pipe>(
-                     index, nodes, diameter.at(index), roughness.at(index),
-                     pipe_status.at(index))));
-      ++index;
+    for (const auto& id : ids) {
+      auto node1id = nodeids[i].first;
+      auto node2id = nodeids[i].second;
+      connected_nodes_.emplace(node1id, nodes_.at(node1id));
+      connected_nodes_.emplace(node2id, nodes_.at(node2id));
+
+      pipes_.emplace(id, std::make_shared<pipenetwork::Pipe>(
+                             id, nodes_.at(node1id), nodes_.at(node2id),
+                             length[i], diameter[i], roughness[i], status[i]));
+      ++i;
     }
+    stat = true;
+
   } catch (std::out_of_range& range_error) {
-    status = false;
-    std::cout << "Pipe(id: " << index
-              << ") is not created, as input node does not exist, nodal id: "
-              << nodeids.at(index).first << " or " << nodeids.at(index).second
-              << '\n'
+    std::cout << "Input node does not exist, nodal id: " << nodeids.at(i).first
+              << " or " << nodeids.at(i).second << '\n'
               << "Pipe creation is unfinished" << '\n';
   }
-  return status;
+  return stat;
```

Comment:
why `stat` means statistics, instead of `status`?

Reply:
I have a variable named status for pipe status (open or close) so I have to use another name for construction status, which is stat....